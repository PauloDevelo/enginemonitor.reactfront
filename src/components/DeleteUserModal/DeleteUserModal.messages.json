{
  "cancel": {
    "defaultMessage": "Cancel",
    "description": "Cancel button",
    "id": "cancel"
  },
  "confirmationMessage1": {
    "defaultMessage": "All the data you input in Equipment Maintenance website will be deleted. It includes your asset, equipments, tasks, and entries and all the images you might have uploaded. ",
    "description": "Confirmation message before deleting the user",
    "id": "deleteUserConfirmationMessage1"
  },
  "confirmationMessage2": {
    "defaultMessage": "After the deletion, you will be logged out from Equipment Maintenance and you won't be able to reconnect.",
    "description": "Confirmation message before deleting the user",
    "id": "deleteUserConfirmationMessage2"
  },
  "confirmationMessage3": {
    "defaultMessage": "If you want to keep proceeding, please type yes below:",
    "description": "Confirmation message before deleting the user",
    "id": "deleteUserConfirmationMessage3"
  },
  "delete": {
    "defaultMessage": "Delete all the user's data",
    "description": "Delete button",
    "id": "deleteUserData"
  },
  "placeHolder": {
    "defaultMessage": "Type yes to confirm the deletion",
    "description": "Place holder",
    "id": "deleteUserConfirmationPlaceHolder"
  },
  "title": {
    "defaultMessage": "Deletion of all the user's data",
    "description": "Title of the delete user modal",
    "id": "deleteUserModalTitle"
  }
}
