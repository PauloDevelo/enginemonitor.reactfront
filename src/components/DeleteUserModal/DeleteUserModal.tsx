import React, { useState, useCallback } from 'react';
import {
  Button, Modal, ModalHeader, ModalBody, ModalFooter, Input,
} from 'reactstrap';

import _ from 'lodash';

import logger from 'loglevel';

import { FormattedMessage, defineMessages, useIntl } from 'react-intl';
import Alerts from '../Alerts/Alerts';
import HttpError from '../../http/HttpError';
import userProxy from '../../services/UserProxy';

import '../../style/transition.css';

import jsonMessages from './DeleteUserModal.messages.json';
import ActionButton from '../ActionButton/ActionButton';

const modalmsg = defineMessages(jsonMessages);

type Props = {
    isOpen: boolean,
    toggle: () => void,
    onUserDeleted: () => void,
    className?: string,
}

const shouldDeletingButtonDisabled = (confirmationMsg: string) => {
  const cleanedConfirmationMsg = _.lowerCase(_.trim(confirmationMsg));
  return cleanedConfirmationMsg !== 'yes' && cleanedConfirmationMsg !== 'oui';
};

const DeleteUserModal = ({
  isOpen,
  toggle,
  onUserDeleted,
  className,
}: Props) => {
  const intl = useIntl();
  const [isActing, setActing] = useState(false);
  const [isDeleteButtonDisabled, setDeleteButtonDisabled] = useState(true);
  const [alerts, setAlerts] = useState<any>(undefined);

  const deleteUser = async () => {
    setActing(true);
    try {
      await userProxy.deleteUser();
      await userProxy.logout();
      toggle();
      onUserDeleted();
      setAlerts(undefined);
      setDeleteButtonDisabled(true);
    } catch (error) {
      if (error instanceof HttpError) {
        setAlerts(error.data);
      } else {
        logger.error(error);
      }
    } finally {
      setActing(false);
    }
  };

  const cancel = () => {
    setAlerts(undefined);
    toggle();
  };

  const onChangeHandler = useCallback((event:React.ChangeEvent<HTMLInputElement>):void => {
    const target:HTMLInputElement = event.target as HTMLInputElement;
    setDeleteButtonDisabled(shouldDeletingButtonDisabled(target.value));
    setAlerts(undefined);
  }, []);

  return (
    <Modal isOpen={isOpen} toggle={toggle} className={className} fade>
      <ModalHeader toggle={toggle}><FormattedMessage {...modalmsg.title} /></ModalHeader>
      <ModalBody>
        <p><FormattedMessage {...modalmsg.confirmationMessage1} /></p>
        <p><FormattedMessage {...modalmsg.confirmationMessage2} /></p>
        <p><FormattedMessage {...modalmsg.confirmationMessage3} /></p>
        <Input placeholder={intl.formatMessage(modalmsg.placeHolder)} onChange={onChangeHandler} />
        <br />
        <Alerts errors={alerts} />
      </ModalBody>
      <ModalFooter>
        <ActionButton isActing={isActing} color="danger" action={deleteUser} message={modalmsg.delete} disabled={isDeleteButtonDisabled} />
        <Button color="secondary" onClick={cancel}><FormattedMessage {...modalmsg.cancel} /></Button>
      </ModalFooter>
    </Modal>
  );
};

export default DeleteUserModal;
