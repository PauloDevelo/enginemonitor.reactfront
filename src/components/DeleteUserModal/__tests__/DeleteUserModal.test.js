/* eslint-disable no-unused-vars */
import chai, { assert } from 'chai';
import React from 'react';
import { IntlProvider } from 'react-intl';
import { mount } from 'enzyme';

import ignoredMessages from '../../../testHelpers/MockConsole';
import updateWrapper from '../../../testHelpers/EnzymeHelper';

import DeleteUserModal from '../DeleteUserModal';

import userProxy from '../../../services/UserProxy';

jest.mock('../../../services/UserProxy');

describe('DeleteUserModal', () => {
  beforeAll(() => {
    ignoredMessages.length = 0;
    ignoredMessages.push('test was not wrapped in act(...)');
    ignoredMessages.push('[React Intl] Missing message');
    ignoredMessages.push('MISSING_TRANSLATION');
  });

  beforeEach(() => {
  });

  afterEach(() => {
    userProxy.deleteUser.mockReset();
    userProxy.logout.mockReset();
  });

  it('Should not call onUserDeleted when loaded', async () => {
    // Arrange
    let deleteUserModalVis = true;

    const toggleDeleteUserModal = () => {
      deleteUserModalVis = !deleteUserModalVis;
    };

    const onUserDeleted = jest.fn();

    // Act
    const deleteUserModal = mount(
      <IntlProvider locale={navigator.language}>
        <DeleteUserModal isOpen={deleteUserModalVis} toggle={toggleDeleteUserModal} onUserDeleted={onUserDeleted} />
      </IntlProvider>,
    );

    // Assert
    expect(onUserDeleted).toHaveBeenCalledTimes(0);
    expect(deleteUserModal).toMatchSnapshot();
  });

  it('Should have a disable button because yes or oui is not in the input.', async () => {
    // Arrange
    let deleteUserModalVis = true;

    const toggleDeleteUserModal = () => {
      deleteUserModalVis = !deleteUserModalVis;
    };

    const onUserDeleted = jest.fn();

    // Act
    const deleteUserModal = mount(
      <IntlProvider locale={navigator.language}>
        <DeleteUserModal isOpen={deleteUserModalVis} toggle={toggleDeleteUserModal} onUserDeleted={onUserDeleted} />
      </IntlProvider>,
    );

    // Assert
    const button = deleteUserModal.find('button.btn-danger').get(0);
    expect(button.props.disabled).toBe(true);
  });

  it('Should have a enable button because yes or oui is in the input.', async () => {
    // Arrange
    let deleteUserModalVis = true;

    const toggleDeleteUserModal = () => {
      deleteUserModalVis = !deleteUserModalVis;
    };

    jest.spyOn(userProxy, 'deleteUser').mockImplementation(() => Promise.resolve());
    jest.spyOn(userProxy, 'logout').mockImplementation(() => Promise.resolve());
    const onUserDeleted = jest.fn();

    const deleteUserModal = mount(
      <IntlProvider locale={navigator.language}>
        <DeleteUserModal isOpen={deleteUserModalVis} toggle={toggleDeleteUserModal} onUserDeleted={onUserDeleted} />
      </IntlProvider>,
    );

    // Act
    const inputs = deleteUserModal.find('input');
    inputs.at(0).simulate('change', { target: { value: 'yes' } });

    // Assert
    const button = deleteUserModal.find('button.btn-danger').get(0);
    expect(button.props.disabled).toBe(false);
  });

  it('Should call onUserDeleted when the user enter yes and validate', async () => {
    // Arrange
    let deleteUserModalVis = true;

    const toggleDeleteUserModal = () => {
      deleteUserModalVis = !deleteUserModalVis;
    };

    const onUserDeleted = jest.fn();

    const deleteUserModal = mount(
      <IntlProvider locale={navigator.language}>
        <DeleteUserModal isOpen={deleteUserModalVis} toggle={toggleDeleteUserModal} onUserDeleted={onUserDeleted} />
      </IntlProvider>,
    );

    const inputs = deleteUserModal.find('input');
    inputs.at(0).simulate('change', { target: { value: 'yes' } });

    const button = deleteUserModal.find('button.btn-danger');

    // Act
    button.simulate('click');
    await updateWrapper(deleteUserModal);

    // Assert
    expect(userProxy.logout).toHaveBeenCalledTimes(1);
    expect(userProxy.deleteUser).toHaveBeenCalledTimes(1);
    expect(onUserDeleted).toHaveBeenCalledTimes(1);
  });
});
