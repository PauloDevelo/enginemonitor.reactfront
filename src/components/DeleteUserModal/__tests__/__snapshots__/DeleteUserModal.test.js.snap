// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteUserModal Should not call onUserDeleted when loaded 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en-US"
  messages={Object {}}
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <DeleteUserModal
    isOpen={true}
    onUserDeleted={[MockFunction]}
    toggle={[Function]}
  >
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      container="body"
      fade={true}
      isOpen={true}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      returnFocusAfterClose={true}
      role="dialog"
      scrollable={false}
      toggle={[Function]}
      trapFocus={false}
      unmountOnClose={true}
      zIndex={1050}
    >
      <Portal
        node={
          <div
            style="position: relative; z-index: 1050; display: block;"
            tabindex="-1"
          >
            <div
              class=""
            >
              <div
                class="modal fade"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog"
                  role="document"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      class="modal-header"
                    >
                      <h5
                        class="modal-title"
                      >
                        Deletion of all the user's data
                      </h5>
                      <button
                        aria-label="Close"
                        class="close"
                        type="button"
                      >
                        <span
                          aria-hidden="true"
                        >
                          ×
                        </span>
                      </button>
                    </div>
                    <div
                      class="modal-body"
                    >
                      <p>
                        All the data you input in Equipment Maintenance website will be deleted. It includes your asset, equipments, tasks, and entries and all the images you might have uploaded. 
                      </p>
                      <p>
                        After the deletion, you will be logged out from Equipment Maintenance and you won't be able to reconnect.
                      </p>
                      <p>
                        If you want to keep proceeding, please type yes below:
                      </p>
                      <input
                        class="form-control"
                        placeholder="Type yes to confirm the deletion"
                        type="text"
                      />
                      <br />
                    </div>
                    <div
                      class="modal-footer"
                    >
                      <button
                        class="btn btn-danger disabled"
                        disabled=""
                        type="button"
                      >
                        Delete all the user's data
                      </button>
                      <button
                        class="btn btn-secondary"
                        type="button"
                      >
                        Cancel
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-backdrop fade"
              />
            </div>
          </div>
        }
      >
        <Portal
          containerInfo={
            <div
              style="position: relative; z-index: 1050; display: block;"
              tabindex="-1"
            >
              <div
                class=""
              >
                <div
                  class="modal fade"
                  role="dialog"
                  style="display: block;"
                  tabindex="-1"
                >
                  <div
                    class="modal-dialog"
                    role="document"
                  >
                    <div
                      class="modal-content"
                    >
                      <div
                        class="modal-header"
                      >
                        <h5
                          class="modal-title"
                        >
                          Deletion of all the user's data
                        </h5>
                        <button
                          aria-label="Close"
                          class="close"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                          >
                            ×
                          </span>
                        </button>
                      </div>
                      <div
                        class="modal-body"
                      >
                        <p>
                          All the data you input in Equipment Maintenance website will be deleted. It includes your asset, equipments, tasks, and entries and all the images you might have uploaded. 
                        </p>
                        <p>
                          After the deletion, you will be logged out from Equipment Maintenance and you won't be able to reconnect.
                        </p>
                        <p>
                          If you want to keep proceeding, please type yes below:
                        </p>
                        <input
                          class="form-control"
                          placeholder="Type yes to confirm the deletion"
                          type="text"
                        />
                        <br />
                      </div>
                      <div
                        class="modal-footer"
                      >
                        <button
                          class="btn btn-danger disabled"
                          disabled=""
                          type="button"
                        >
                          Delete all the user's data
                        </button>
                        <button
                          class="btn btn-secondary"
                          type="button"
                        >
                          Cancel
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="modal-backdrop fade"
                />
              </div>
            </div>
          }
        >
          <div
            className=""
          >
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onClick={[Function]}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              role="dialog"
              style={
                Object {
                  "display": "block",
                }
              }
              tabIndex="-1"
              tag="div"
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div
                  className="modal fade"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                  tabIndex="-1"
                >
                  <div
                    className="modal-dialog"
                    role="document"
                  >
                    <div
                      className="modal-content"
                    >
                      <ModalHeader
                        charCode={215}
                        closeAriaLabel="Close"
                        tag="h5"
                        toggle={[Function]}
                        wrapTag="div"
                      >
                        <div
                          className="modal-header"
                        >
                          <h5
                            className="modal-title"
                          >
                            <FormattedMessage
                              defaultMessage="Deletion of all the user's data"
                              description="Title of the delete user modal"
                              id="deleteUserModalTitle"
                              values={Object {}}
                            >
                              Deletion of all the user's data
                            </FormattedMessage>
                          </h5>
                          <button
                            aria-label="Close"
                            className="close"
                            onClick={[Function]}
                            type="button"
                          >
                            <span
                              aria-hidden="true"
                            >
                              ×
                            </span>
                          </button>
                        </div>
                      </ModalHeader>
                      <ModalBody
                        tag="div"
                      >
                        <div
                          className="modal-body"
                        >
                          <p>
                            <FormattedMessage
                              defaultMessage="All the data you input in Equipment Maintenance website will be deleted. It includes your asset, equipments, tasks, and entries and all the images you might have uploaded. "
                              description="Confirmation message before deleting the user"
                              id="deleteUserConfirmationMessage1"
                              values={Object {}}
                            >
                              All the data you input in Equipment Maintenance website will be deleted. It includes your asset, equipments, tasks, and entries and all the images you might have uploaded. 
                            </FormattedMessage>
                          </p>
                          <p>
                            <FormattedMessage
                              defaultMessage="After the deletion, you will be logged out from Equipment Maintenance and you won't be able to reconnect."
                              description="Confirmation message before deleting the user"
                              id="deleteUserConfirmationMessage2"
                              values={Object {}}
                            >
                              After the deletion, you will be logged out from Equipment Maintenance and you won't be able to reconnect.
                            </FormattedMessage>
                          </p>
                          <p>
                            <FormattedMessage
                              defaultMessage="If you want to keep proceeding, please type yes below:"
                              description="Confirmation message before deleting the user"
                              id="deleteUserConfirmationMessage3"
                              values={Object {}}
                            >
                              If you want to keep proceeding, please type yes below:
                            </FormattedMessage>
                          </p>
                          <Input
                            onChange={[Function]}
                            placeholder="Type yes to confirm the deletion"
                            type="text"
                          >
                            <input
                              className="form-control"
                              onChange={[Function]}
                              placeholder="Type yes to confirm the deletion"
                              type="text"
                            />
                          </Input>
                          <br />
                          <Alerts />
                        </div>
                      </ModalBody>
                      <ModalFooter
                        tag="div"
                      >
                        <div
                          className="modal-footer"
                        >
                          <ActionButton
                            action={[Function]}
                            color="danger"
                            disabled={true}
                            isActing={false}
                            message={
                              Object {
                                "defaultMessage": "Delete all the user's data",
                                "description": "Delete button",
                                "id": "deleteUserData",
                              }
                            }
                          >
                            <Button
                              color="danger"
                              disabled={true}
                              onClick={[Function]}
                              tag="button"
                              type="button"
                            >
                              <button
                                aria-label={null}
                                className="btn btn-danger disabled"
                                disabled={true}
                                onClick={[Function]}
                                type="button"
                              >
                                <FormattedMessage
                                  defaultMessage="Delete all the user's data"
                                  description="Delete button"
                                  id="deleteUserData"
                                  values={Object {}}
                                >
                                  Delete all the user's data
                                </FormattedMessage>
                              </button>
                            </Button>
                          </ActionButton>
                          <Button
                            color="secondary"
                            onClick={[Function]}
                            tag="button"
                          >
                            <button
                              aria-label={null}
                              className="btn btn-secondary"
                              onClick={[Function]}
                              type="button"
                            >
                              <FormattedMessage
                                defaultMessage="Cancel"
                                description="Cancel button"
                                id="cancel"
                                values={Object {}}
                              >
                                Cancel
                              </FormattedMessage>
                            </button>
                          </Button>
                        </div>
                      </ModalFooter>
                    </div>
                  </div>
                </div>
              </Transition>
            </Fade>
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal-backdrop"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={true}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              tag="div"
              timeout={150}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={150}
                unmountOnExit={false}
              >
                <div
                  className="modal-backdrop fade"
                />
              </Transition>
            </Fade>
          </div>
        </Portal>
      </Portal>
    </Modal>
  </DeleteUserModal>
</IntlProvider>
`;
